# Agents in Autogen 

In [5]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_agentchat.messages import TextMessage
from dotenv import load_dotenv
import os

load_dotenv()
api_key = os.getenv('GEMINI_API_KEY')
model_client = OpenAIChatCompletionClient(model='gemini-1.5-flash-8b', api_key=api_key)

In [6]:
async def web_search(query: str) -> str:
    """Find information on the web"""
    return "The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. "

In [7]:
agent = AssistantAgent(
    name='assistant',
    model_client=model_client,
    tools=[web_search],
    system_message='Use Tools to solve tasks',
    description='An agent which uses tool to help solve task'
)

In [8]:
result = await agent.run(task='Find information about Labrador Retriever')
print(result.messages[-1].content)

The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. 


In [9]:
result

TaskResult(messages=[TextMessage(id='f636a702-8c33-4474-8b23-47778785044e', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 8, 27, 10, 9, 44, 624626, tzinfo=datetime.timezone.utc), content='Find information about Labrador Retriever', type='TextMessage'), ToolCallRequestEvent(id='fb29f1d1-7f94-4653-9b05-55c0eddebe14', source='assistant', models_usage=RequestUsage(prompt_tokens=23, completion_tokens=7), metadata={}, created_at=datetime.datetime(2025, 8, 27, 10, 9, 45, 478715, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='', arguments='{"query":"Labrador Retriever"}', name='web_search')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='8829bab2-20b6-4b69-8f9f-a7357ba978f7', source='assistant', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 8, 27, 10, 9, 45, 479559, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun do

TaskResult(messages=
[
    1. TextMessage(source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 6, 19, 16, 0, 26, 684759, tzinfo=datetime.timezone.utc), content='Find information about Labrador Retriever', type='TextMessage')
    
    2. ToolCallRequestEvent(source='assistant', models_usage=RequestUsage(prompt_tokens=61, completion_tokens=17), metadata={}, created_at=datetime.datetime(2025, 6, 19, 16, 0, 27, 992815, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_ipWxcrV4tXyUsapUhqpz6NyO', arguments='{"query":"Labrador Retriever"}', name='web_search')], type='ToolCallRequestEvent'), 
    
    3. ToolCallExecutionEvent(source='assistant', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 6, 19, 16, 0, 27, 993741, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', name='web_search', call_id='call_ipWxcrV4tXyUsapUhqpz6NyO', is_error=False)], type='ToolCallExecutionEvent'), 
    
    4. ToolCallSummaryMessage(source='assistant', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 6, 19, 16, 0, 27, 993775, tzinfo=datetime.timezone.utc), content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', type='ToolCallSummaryMessage', tool_calls=[FunctionCall(id='call_ipWxcrV4tXyUsapUhqpz6NyO', arguments='{"query":"Labrador Retriever"}', name='web_search')], 
    
    results=[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', name='web_search', call_id='call_ipWxcrV4tXyUsapUhqpz6NyO', is_error=False)])], stop_reason=None)

### On_message() method

In [10]:
from autogen_core import CancellationToken

async def assistant_run()-> None:
    response = await agent.on_messages(
        messages= [TextMessage(content='Find information about Labrador Retriever',source='User')],
        cancellation_token=CancellationToken()
    )

    print(response.inner_messages)
    print('\n\n\n\n')
    print(response.chat_message)

await assistant_run()

[ToolCallRequestEvent(id='95a73f0b-61d8-4dc6-933b-4635db92bca6', source='assistant', models_usage=RequestUsage(prompt_tokens=55, completion_tokens=7), metadata={}, created_at=datetime.datetime(2025, 8, 27, 10, 13, 12, 900955, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='', arguments='{"query":"Labrador Retriever"}', name='web_search')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='cb0db424-2e31-4685-87e8-fbdc384b800f', source='assistant', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 8, 27, 10, 13, 12, 901636, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', name='web_search', call_id='', is_error=False)], type='ToolCallExecutionEvent')]





id='3b27a124-d84a-4737-a0a9-197c4c77985e' source='assistant' models_usage=None metadata={} created_at=datetime.datetime(2025, 8, 27, 10, 13, 12, 901682, tzinfo=datetime.timezone.utc) content='Th

[ToolCallRequestEvent(id='95a73f0b-61d8-4dc6-933b-4635db92bca6', source='assistant', models_usage=RequestUsage(prompt_tokens=55, completion_tokens=7), metadata={}, created_at=datetime.datetime(2025, 8, 27, 10, 13, 12, 900955, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='', arguments='{"query":"Labrador Retriever"}', name='web_search')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='cb0db424-2e31-4685-87e8-fbdc384b800f', source='assistant', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 8, 27, 10, 13, 12, 901636, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', name='web_search', call_id='', is_error=False)], type='ToolCallExecutionEvent')]





id='3b27a124-d84a-4737-a0a9-197c4c77985e' source='assistant' models_usage=None metadata={} created_at=datetime.datetime(2025, 8, 27, 10, 13, 12, 901682, tzinfo=datetime.timezone.utc) content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ' type='ToolCallSummaryMessage' tool_calls=[FunctionCall(id='', arguments='{"query":"Labrador Retriever"}', name='web_search')] results=[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', name='web_search', call_id='', is_error=False)]

source='assistant' models_usage=None metadata={} created_at=datetime.datetime(2025, 6, 19, 16, 3, 51, 196163, tzinfo=datetime.timezone.utc) content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ' type='ToolCallSummaryMessage' tool_calls=[FunctionCall(id='call_WYA4GSiwmxw7lOBDMmf5jHqE', arguments='{"query":"Labrador Retriever breed information"}', name='web_search')] results=[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', name='web_search', call_id='call_WYA4GSiwmxw7lOBDMmf5jHqE', is_error=False)]

### Streaming Messages

#### on_messages_stream()

In [11]:
from autogen_agentchat.ui import Console


async def assistant_run_stream() -> None:

    await Console(
        agent.on_messages_stream(
        messages= [TextMessage(content='Find information about Labrador Retriever via the tool',source='User')],
        cancellation_token=CancellationToken()
    ),
    output_stats=True # Enable stats Printing
    )


await assistant_run_stream()

---------- ToolCallRequestEvent (assistant) ----------
[FunctionCall(id='', arguments='{"query":"Labrador Retriever"}', name='web_search')]
[Prompt tokens: 90, Completion tokens: 7]
---------- ToolCallExecutionEvent (assistant) ----------
[FunctionExecutionResult(content='The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. ', name='web_search', call_id='', is_error=False)]
---------- assistant ----------
The Labrador Retriever or simply Labrador is a British breed of retriever gun dog. 
---------- Summary ----------
Number of inner messages: 2
Total prompt tokens: 90
Total completion tokens: 7
Duration: 0.82 seconds


In [12]:
from autogen_agentchat.ui import Console


async def assistant_run_stream() -> None:

    await Console(
        agent.on_messages_stream(
        messages= [TextMessage(content='What was the last question I asked ?',source='User')],
        cancellation_token=CancellationToken()
    ),
    output_stats=True # Enable stats Printing
    )


await assistant_run_stream()

---------- assistant ----------
Your last question was: "Find information about Labrador Retriever via the tool".

[Prompt tokens: 125, Completion tokens: 16]
---------- Summary ----------
Number of inner messages: 0
Total prompt tokens: 125
Total completion tokens: 16
Duration: 1.24 seconds
